### PR TITLE
[ramda] Use mapped types for pluck, prop and props

### LINKED ISSUE #21238: @types/ramda definitions break under Typescript 2.6.1
- [x] I tried using the `@types/ramda` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @donnut, @mdekrey, @mrdziuban, @sbking, @afharo, @teves-castro, @1M0reBug, @hojberg, @charlespwd, @samsonkeung

------

Using the latest Typescript compiler (2.6.1) the following code breaks because of the definitions, this did not happen on 2.5.3
```ts
import { compose, pluck, product, props, sum } from 'ramda'

export interface IItemLineForTotals {
  qty: number
  price: number
}

export const getItemTotal: (itemLine: IItemLineForTotals) => number = compose(
  product,
  props(['qty', 'price']),
)
export const getItemsQty: (itemLines: IItemLineForTotals[]) => number = compose(sum, pluck('qty'))
```

```
ERROR in [at-loader] ./src/ui/lib/itemLineHelpers.ts:9:3 
    TS2345: Argument of type '(list: number[]) => number' is not assignable to parameter of type '(x: {}[]) => number'.
  Types of parameters 'list' and 'x' are incompatible.
    Type '{}[]' is not assignable to type 'number[]'.
      Type '{}' is not assignable to type 'number'.

ERROR in [at-loader] ./src/ui/lib/itemLineHelpers.ts:12:81 
    TS2345: Argument of type '(list: number[]) => number' is not assignable to parameter of type '(x: {}[]) => number'.
  Types of parameters 'list' and 'x' are incompatible.
    Type '{}[]' is not assignable to type 'number[]'.
```

### PR DESCRIPTION
**Pull request by pelotom**
Resolves #21238.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

File: types/ramda/ramda-tests.ts
Comment (pelotom) [inline types/ramda/ramda-tests.ts]: This shouldn't type check, so it shouldn't be a test case.

File: types/ramda/ramda-tests.ts
Comment (pelotom) [inline types/ramda/ramda-tests.ts]: This shouldn't type check, so it shouldn't be a test case.

File: types/ramda/ramda-tests.ts
Comment (pelotom) [inline types/ramda/ramda-tests.ts]: This shouldn't type check, so it shouldn't be a test case.

File: types/ramda/ramda-tests.ts
Comment (pelotom) [inline types/ramda/ramda-tests.ts]: The type parameters to `compose` were wrong here, which came to light as a result of these improved typings.

**Review by samsonkeung — APPROVED** [review]

**Comment by RyanCavanaugh** [comment]
Approved by a listed owner. PR ready to merge pending express review by a maintainer.

File: types/ramda/index.d.ts
Comment (DanielRosenwasser) [inline types/ramda/index.d.ts]: What do you think of the following type instead?

```ts
prop<P extends string>(p: P): <T extends Record<P, any>(obj: T): Pick<T, P>;
```

Comment (pelotom) [reply types/ramda/index.d.ts]: AFAIK the they’re equivalent; is there a reason to prefer that? I find yours a little more complex to read.

Comment (pelotom) [reply types/ramda/index.d.ts]: Wait, you mean `O[P]` instead of `Pick<T, P>`, don’t you?

Comment (DanielRosenwasser) [reply types/ramda/index.d.ts]: Oh, yes, probably. Sorry, I don't use it that much.

Comment (DanielRosenwasser) [reply types/ramda/index.d.ts]: They're actually not exactly equivalent. With `Record<P, T>`, you're expecting an object that is homogeneous for some well-known `T` (meaning that if you have `{x: number, y: string}`, you will always get `number | string`. If you make `T` the object type, and then use an indexed access, you've delayed the evaluation of the return type until `T` is fully known, giving you a more accurate `prop`.

Comment (pelotom) [reply types/ramda/index.d.ts]: @DanielRosenwasser ok, but in this case `P` can't be a union of multiple literals, since it's determined by a single `p: P`. So `Record<P, T>` amounts to just `{ [p]: T }`.

Comment (pelotom) [reply types/ramda/index.d.ts]: Well, I guess you could always explicitly widen `P` to a union of literals or even `string`. Fair point, I think you're right.

**Pull request merged**
